dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java

```diff
@@ -0,0 +1,400 @@
+	{
+		this.fromAddress = toInternetAddress(fromAddress).orElse(null);
+		this.toAddresses = toAddresses == null ? new InternetAddress[0]
+				: toAddresses.stream().flatMap(s -> toInternetAddress(s).stream()).toArray(InternetAddress[]::new);
+		this.toAddressesCc = toAddressesCc == null ? new InternetAddress[0]
+				: toAddressesCc.stream().flatMap(s -> toInternetAddress(s).stream()).toArray(InternetAddress[]::new);
+		this.replyToAddresses = replyToAddresses == null ? new InternetAddress[0]
+				: replyToAddresses.stream().flatMap(s -> toInternetAddress(s).stream()).toArray(InternetAddress[]::new);
+
+		session = createSession(useSmtps, mailServerHostname, mailServerPort, mailServerUsername, mailServerPassword,
+				trustStore, keyStore, keyStorePassword);
+
+		smimeSignedGenerator = createSmimeSignedGenerator(fromAddress, signStore, signStorePassword);
+	}
+
+	private Optional<InternetAddress> toInternetAddress(String fromAddress)
+	{
+		if (fromAddress == null || fromAddress.isBlank())
+			return Optional.empty();
+
+		try
+		{
+			return Optional.of(new InternetAddress(fromAddress));
+		}
+		catch (AddressException e)
+		{
+			logger.warn("Unable to create {} from {}: {} - {}", InternetAddress.class.getName(), fromAddress,
+					e.getClass().getName(), e.getMessage());
+
+			return Optional.empty();
+		}
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception
+	{
+		if (fromAddress == null)
+			throw new IllegalArgumentException("no valid from address configured");
+		if (toAddresses.length == 0)
+			throw new IllegalArgumentException("no valid to addresses configured");
+	}
+
+	private Session createSession(boolean useSmtps, String mailServerHostname, int mailServerPort,
+			String mailServerUsername, char[] mailServerPassword, KeyStore trustStore, KeyStore keyStore,
+			char[] keyStorePassword)
+	{
+		Properties properties = new Properties();
+
+		Authenticator authenticator = null;
+		if (mailServerUsername != null && mailServerPassword != null)
+		{
+			authenticator = new Authenticator()
+			{
+				@Override
+				protected PasswordAuthentication getPasswordAuthentication()
+				{
+					return new PasswordAuthentication(mailServerUsername, String.copyValueOf(mailServerPassword));
+				}
+			};
+
+			properties.put("mail.smtp.auth", "true");
+		}
+
+		if (useSmtps)
+		{
+			properties.put("mail.smtp.ssl.enable", "true");
+			properties.put("mail.smtp.host", mailServerHostname);
+			properties.put("mail.smtp.port", mailServerPort);
+
+			properties.put("mail.transport.protocol", "smtps");
+			properties.put("mail.smtp.socketFactory.fallback", "false");
+
+			properties.put("mail.smtp.ssl.checkserveridentity", "true");
+			properties.put("mail.smtp.ssl.socketFactory",
+					createSslSocketFactory(trustStore, keyStore, keyStorePassword));
+		}
+		else
+		{
+			properties.put("mail.smtp.host", mailServerHostname);
+			properties.put("mail.smtp.port", mailServerPort);
```

Comment:
property set twice (L186/L198), could be moved to the top where the properties are initialized

Reply:
fixed with bdb79ce